### PR TITLE
make models/phi3.py and models/phi3small.py compatible with mypy

### PR DESCRIPTION
**Pull request by wangkuiyi**
## Test Plan:

Run the following command:

```shell
❯ mypy \
  --explicit-package-bases \
  --ignore-missing-imports \
. | grep models/phi
```

Before this change, it should complain the following:

```
mlx_lm/models/phi3small.py:22: error: Incompatible types in assignment (expression has type "None", variable has type "int")  [assignment]
mlx_lm/models/phi3small.py:29: error: Incompatible types in assignment (expression has type "Tuple[int]", variable has type "int")  [assignment]
mlx_lm/models/phi3small.py:74: error: Name "math" is not defined  [name-defined]
mlx_lm/models/phi3small.py:176: error: "Tuple[Any, Any]" has no attribute "offset"  [attr-defined]
mlx_lm/models/phi3small.py:177: error: "Tuple[Any, Any]" has no attribute "offset"  [attr-defined]
mlx_lm/models/phi3small.py:178: error: "Tuple[Any, Any]" has no attribute "update_and_fetch"  [attr-defined]
mlx_lm/models/phi3.py:20: error: Incompatible types in assignment (expression has type "None", variable has type "int")  [assignment]
mlx_lm/models/phi3.py:68: error: Argument "short_factor" to "SuScaledRotaryEmbedding" has incompatible type "Union[float, str]"; expected "Union[List[float], float]"  [arg-type]
mlx_lm/models/phi3.py:69: error: Argument "long_factor" to "SuScaledRotaryEmbedding" has incompatible type "Union[float, str]"; expected "Union[List[float], float]"  [arg-type]
mlx_lm/models/phi3.py:73: error: Unsupported operand types for / ("int" and "str")  [operator]
mlx_lm/models/phi3.py:73: note: Right operand is of type "Union[float, str]"
mlx_lm/models/phi3.py:101: error: "Tuple[Any, Any]" has no attribute "offset"  [attr-defined]
mlx_lm/models/phi3.py:102: error: "Tuple[Any, Any]" has no attribute "offset"  [attr-defined]
mlx_lm/models/phi3.py:103: error: "Tuple[Any, Any]" has no attribute "update_and_fetch"  [attr-defined]
```

After this change, the above command should complain no more.

### CODE REVIEW

File: llms/mlx_lm/models/phi3.py [hunk 1]
```diff
@@ -17,10 +17,10 @@ class ModelArgs(BaseModelArgs):
     num_attention_heads: int
     rms_norm_eps: float
     vocab_size: int
-    num_key_value_heads: int = None
+    num_key_value_heads: Optional[int] = None
     rope_theta: float = 10000
     rope_traditional: bool = False
-    rope_scaling: Optional[Dict[str, Union[float, str]]] = None
+    rope_scaling: Optional[Dict[str, Union[float, List[float]]]] = None
```
Comment (wangkuiyi) [inline llms/mlx_lm/models/phi3.py:23]: Using Python dictionary implicitly presumed schemeless or typeless coding style, which is against the style of the rest fields in this dataclass. We may want to change this dictionary into a nested dataclass in the future.

Comment (awni) [reply llms/mlx_lm/models/phi3.py:23]: Indeed that would probably be better.

**Review by awni — APPROVED** [review]
🙏  thanks for the fixes. Just letting the tests clear then I will merge it.

File: llms/mlx_lm/models/phi3.py (hunk 1 above)
Comment (wangkuiyi) [reply llms/mlx_lm/models/phi3.py:23]: I will try if that works today, then file another PR if it does.

**Pull request merged**